Add Fast Segformer Processor

capnmav77 opened the pull request:
# What does this PR do?
Linked: https://github.com/huggingface/transformers/issues/36978

Adding fast processor for Segformer.

CC: @yonigozlan
<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->


<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker
- vision models: @amyeroberts, @qubvel
- speech models: @eustlb
- graph models: @clefourrier

Library:

- flax: @gante and @Rocketknight1
- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- pipelines: @Rocketknight1
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @SunMarc
- chat templates: @Rocketknight1

Integrations:

- deepspeed: HF Trainer/Accelerate: @muellerzr
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc @MekkCyber

Documentation: @stevhliu

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @Rocketknight1
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

capnmav77 committed:
Add Fast Segformer Processor

capnmav77 committed:
Modified the params according to segformer model

capnmav77 committed:
modified test_image_processing_Segformer_fast args

- removed redundant params like do_center_crop,center_crop which aren't present in the original segformer class

yonigozlan reviewed (commented):
Hi @capnmav77 , Thanks for working on this! Looks like the `segmentation_maps` processing logic from the slow image processor is missing here. You can take a look at how it's being handled for `BeitImageProcessorFast` as it is very similar to Segformer image processor! Here is the PR `BeitImageProcessorFast` https://github.com/huggingface/transformers/pull/37005

capnmav77 committed:
added segmentation_maps processing logic form the slow segformer processing module with references from beitimageprocessing fast

capnmav77 commented:
> Hi @capnmav77 , Thanks for working on this! Looks like the `segmentation_maps` processing logic from the slow image processor is missing here. You can take a look at how it's being handled for `BeitImageProcessorFast` as it is very similar to Segformer image processor! Here is the PR `BeitImageProcessorFast` #37005

hey i've added the segmentation_maps processing logic from the slow image processor also i've taken references from the beitimageprocessorFast it really helped a lot , do let me know what to do next

capnmav77 committed:
Merge branch 'main' into main

capnmav77 committed:
fixed code_quality

yonigozlan commented on src/transformers/models/segformer/image_processing_segformer_fast.py:
```diff
@@ -0,0 +1,337 @@
+        prepared_segmentation_maps = None
+        if segmentation_maps is not None:
+            # Segmentation maps should not be converted to RGB
+            prepared_segmentation_maps = self._prepare_input_images(
+                images=segmentation_maps,
+                do_convert_rgb=False, # Explicitly false for masks
+                input_data_format=input_data_format, # Detect format initially
+                device=device,
+            )
```
This may cause issues, see my comment in Beit: https://github.com/huggingface/transformers/pull/37005#discussion_r2017411063

Maybe the best thing to do here is to add a `self._prepare_input_segmentation_maps` method

yonigozlan commented on src/transformers/models/segformer/image_processing_segformer_fast.py:
```diff
@@ -0,0 +1,337 @@
+
+        return BatchFeature(data=data, tensor_type=kwargs.get("return_tensors"))
+
+    # Copied from transformers.models.beit.image_processing_beit_fast.BeitImageProcessorFast.post_process_semantic_segmentation with Beit->Segformer
```
This will cause issues with the CI right now because `BeitImageProcessorFast` doesn't exist yet. I would just remove it for now.

yonigozlan commented on tests/models/segformer/test_image_processing_segformer.py:
```diff
@@ -120,13 +124,14 @@ def image_processor_dict(self):
-        self.assertTrue(hasattr(image_processing, "image_mean"))
-        self.assertTrue(hasattr(image_processing, "image_std"))
-        self.assertTrue(hasattr(image_processing, "do_reduce_labels"))
+        for image_processing_class in self.image_processor_list:
+            image_processing = image_processing_class(**self.image_processor_dict)
```
The same should be done for all other tests!

yonigozlan reviewed (commented):
Hey @capnmav77 thanks a lot for iterating, looks much better! Still a few things to change, and now the important part is to check that all tests pass with 
```bash
RUN_SLOW=1 python -m pytest tests/models/segformer/test_image_processing_segformer.py
```

You should also override `test_slow_fast_equivalence` to also test segmentation maps processing. You can check how it was done in the Beit PR.

Also all the formatting/repo consistency tests should pass. You can use the utilities `make fixup`, `make fix-copies`, `make style`, `make quality` to help.

capnmav77 committed:
added recommended fixes and tests to make sure everything processess smoothly

capnmav77 commented:
Hey @yonigozlan , i have modified the code accordingly but there's a issue i'm not able to figure out , one local test fails : 
```
FAILED tests/models/segformer/test_image_processing_segformer.py::SegformerImageProcessingTest::test_slow_fast_equivalence - AssertionError: False is not true
```
do you have a fix for this i'm not able to understand why this is happening , the dims of the fast processor are 1 , C , H , W , which shouldn't be the case compared to the slow processor which has 1, H , W , i presume the labels are messed up somewhere . Please do lemme know the fix for that and do suggest code changes . 

here's the log output of the labels : 
```
tensor([[[ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,
           3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,
           3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,
           3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,
           3,  3,  2,  2,  3,  3,  3,  3,  3,  3,  3,  3,  3],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,
           3,  3,  2,  2,  3,  3,  3,  3,  3,  3,  3,  3,  3],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,
           3,  2,  2,  2,  3,  3,  3,  3,  3,  3,  3,  3,  3],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,
           2,  2,  2,  2,  3,  3,  3,  3,  3,  3,  3,  3,  3],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  2,  2,  3,  3,  2,  2,
           2,  2,  2,  2,  3,  3,  3,  3,  3,  3,  3,  3,  3],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  2,  2,  3,  2,  2,  2,
           2,  2,  2,  2,  3,  3,  3,  3,  3,  3,  3,  3,  3],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,
           2,  2,  2,  2,  3,  3,  3,  3,  3,  3,  3,  3,  3],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,
           2,  2,  2,  2,  2,  3,  3,  3,  3,  3,  3,  2,  2],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,  2,  2,  2,
           2,  2,  2,  2,  2,  3,  3,  3,  3,  2,  2,  2,  2],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,  2,  2,  2,
           2,  2,  2,  2,  2,  3,  2,  2,  2,  2,  2,  2,  2],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,  2,  2,  2,
           2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,  2,  2,  2,
           2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,  2,  2,  2,
           2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2],
         [ 3,  3,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,
           2,  2,  2,  2,  2,  2,  2, 18, 18,  2,  2,  2,  2],
         [ 3,  3,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,
           2,  2,  2,  2,  2,  2,  2, 18,  2,  2,  2,  2,  2],
         [ 5,  5,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,
           2,  2,  2,  2,  2,  2,  2, 18,  2,  2,  2,  2,  2],
         [ 5,  5,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,
           2,  2,  2,  2,  2,  2,  2, 18,  2,  2,  2,  2,  2],
         [ 5,  5,  5,  5,  5,  5,  5,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,
           2,  2,  2,  2,  2,  2,  2, 18, 18,  2,  2,  2,  2],
         [ 5,  5,  5,  5,  5,  5,  5,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,
           2,  2,  2,  2,  2,  2, 18, 18, 18,  2,  2,  2,  2],
         [ 5,  5,  5,  5,  5,  5,  5,  1,  1,  0,  0,  0,  2,  2,  2,  2,  2,
           2,  2,  2,  2,  2,  2, 18, 18, 18,  2,  2,  2,  2],
         [ 1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  2,  2,
           2,  2,  2,  2,  2,  2, 18, 18, 18,  2,  2,  2,  2],
         [ 1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  2,  2,
          10, 10,  2,  2,  2, 18, 18, 18, 18, 18, 18, 18,  2],
         [10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,
          10, 10, 10, 10, 10, 10, 10, 18, 18, 18, 18, 18, 18],
         [10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,
          10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10],
         [10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,
          10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10],
         [10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,
          10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10],
         [10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,
          10, 10, 10, 10, 10, 10, 10, 10,  7,  7,  7,  7,  7]]])
tensor([[[[ 86,  87,  86,  ...,  82,  81,  80],
          [ 89,  90,  90,  ...,  83,  82,  80],
          [ 90,  92,  93,  ...,  85,  84,  83],
          ...,
          [ 99, 100, 101,  ..., 105, 101,  99],
          [102, 107, 107,  ..., 110, 108, 104],
          [ 97,  97, 103,  ..., 121, 119, 117]],

         [[138, 138, 138,  ..., 132, 131, 131],
          [140, 141, 141,  ..., 133, 132, 131],
          [142, 143, 143,  ..., 135, 134, 132],
          ...,
          [109, 108, 108,  ..., 108, 103, 102],
          [107, 108, 105,  ..., 111, 108, 106],
          [102,  99, 102,  ..., 114, 114, 114]],

         [[191, 193, 194,  ..., 191, 189, 188],
          [194, 196, 197,  ..., 194, 191, 189],
          [196, 199, 200,  ..., 196, 193, 191],
          ...,
          [ 62,  63,  63,  ...,  66,  64,  63],
          [ 62,  66,  66,  ...,  71,  70,  71],
          [ 58,  58,  61,  ..., 100, 102, 102]]]])
          ```

yonigozlan commented on src/transformers/models/segformer/image_processing_segformer_fast.py:
```diff
@@ -0,0 +1,301 @@
+        if segmentation_maps is not None:
+            # Segmentation maps should not be converted to RGB
+            prepared_segmentation_maps = make_list_of_images(
+                images = prepared_images,
```
This should be segmentation_maps no? Also the segmentation maps should be converted to torch tensors first. You can use a similar logic as the one in the slow processor for converting segmentation maps to numpy arrays, including the `added_channel_dim` logic

capnmav77 replied:
yes i've shifted the logic to _preprocess_segmentation_maps do give a look

yonigozlan commented on src/transformers/models/segformer/image_processing_segformer_fast.py:
```diff
@@ -0,0 +1,301 @@
+                segmentation_maps=prepared_segmentation_maps,
+                **kwargs,
+            )
+            data["labels"] = processed_segmentation_maps
```
Here you'll need to take into account `added_channel_dim` to squeeze or not the channel dim, as in the slow processor

yonigozlan commented on src/transformers/models/segformer/image_processing_segformer_fast.py:
```diff
@@ -0,0 +1,301 @@
+    do_reduce_labels = False
+    valid_kwargs = SegformerFastImageProcessorKwargs
+
+    def __init__(self, do_reduce_labels=None, **kwargs: Unpack[SegformerFastImageProcessorKwargs]):
```
We should keep do_reduce_labels in the kwargs:
```suggestion
    def __init__(self, **kwargs: Unpack[SegformerFastImageProcessorKwargs]):
```

yonigozlan reviewed (commented):
Hi @capnmav77 thanks for iterating! Sorry for the delayed review. Almost ready to go, although some logic is missing for segmentation maps

capnmav77 committed:
Fixed SegmentationMapsLogic

- modified the preprocessing of segmentation maps to use tensors
- added batch support

capnmav77 commented:
Hi, @yonigozlan  , all the tests pass apart from the following : 
FAILED tests/models/segformer/test_image_processing_segformer.py::SegformerImageProcessingTest::test_slow_fast_equivalence - AssertionError: False is not true

the tensors from the fast processor does not match the tensor from the slow one ? i've been trying to crack this issue but i'm not getting to the root cause , i assume it's some mishap in the reduce_label function ? or something related to labelling can you suggest fixes ? 
```
slow: 
tensor([[[ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,
           3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,
           3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,
           3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,
           3,  3,  2,  2,  3,  3,  3,  3,  3,  3,  3,  3,  3],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,
           3,  3,  2,  2,  3,  3,  3,  3,  3,  3,  3,  3,  3],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,
           3,  2,  2,  2,  3,  3,  3,  3,  3,  3,  3,  3,  3],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,
           2,  2,  2,  2,  3,  3,  3,  3,  3,  3,  3,  3,  3],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  2,  2,  3,  3,  2,  2,
           2,  2,  2,  2,  3,  3,  3,  3,  3,  3,  3,  3,  3],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  2,  2,  3,  2,  2,  2,
           2,  2,  2,  2,  3,  3,  3,  3,  3,  3,  3,  3,  3],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,
           2,  2,  2,  2,  3,  3,  3,  3,  3,  3,  3,  3,  3],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,
           2,  2,  2,  2,  2,  3,  3,  3,  3,  3,  3,  2,  2],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,  2,  2,  2,
           2,  2,  2,  2,  2,  3,  3,  3,  3,  2,  2,  2,  2],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,  2,  2,  2,
           2,  2,  2,  2,  2,  3,  2,  2,  2,  2,  2,  2,  2],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,  2,  2,  2,
           2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,  2,  2,  2,
           2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2],
         [ 3,  3,  3,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,  2,  2,  2,
           2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2],
         [ 3,  3,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,
           2,  2,  2,  2,  2,  2,  2, 18, 18,  2,  2,  2,  2],
         [ 3,  3,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,
           2,  2,  2,  2,  2,  2,  2, 18,  2,  2,  2,  2,  2],
         [ 5,  5,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,
           2,  2,  2,  2,  2,  2,  2, 18,  2,  2,  2,  2,  2],
         [ 5,  5,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,
           2,  2,  2,  2,  2,  2,  2, 18,  2,  2,  2,  2,  2],
         [ 5,  5,  5,  5,  5,  5,  5,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,
           2,  2,  2,  2,  2,  2,  2, 18, 18,  2,  2,  2,  2],
         [ 5,  5,  5,  5,  5,  5,  5,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,
           2,  2,  2,  2,  2,  2, 18, 18, 18,  2,  2,  2,  2],
         [ 5,  5,  5,  5,  5,  5,  5,  1,  1,  0,  0,  0,  2,  2,  2,  2,  2,
           2,  2,  2,  2,  2,  2, 18, 18, 18,  2,  2,  2,  2],
         [ 1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  2,  2,
           2,  2,  2,  2,  2,  2, 18, 18, 18,  2,  2,  2,  2],
         [ 1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  2,  2,
          10, 10,  2,  2,  2, 18, 18, 18, 18, 18, 18, 18,  2],
         [10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,
          10, 10, 10, 10, 10, 10, 10, 18, 18, 18, 18, 18, 18],
         [10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,
          10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10],
         [10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,
          10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10],
         [10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,
          10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10],
         [10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,
          10, 10, 10, 10, 10, 10, 10, 10,  7,  7,  7,  7,  7 ]]]) - here is the change
 Fast: 
tensor([[[[ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,
            3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3],
          [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,
            3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3],
          [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,
            3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3],
          [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,
            3,  3,  2,  2,  3,  3,  3,  3,  3,  3,  3,  3,  3],
          [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,
            3,  2,  2,  2,  3,  3,  3,  3,  3,  3,  3,  3,  3],
          [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,
            3,  2,  2,  2,  3,  3,  3,  3,  3,  3,  3,  3,  3],
          [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  2,
            2,  2,  2,  2,  3,  3,  3,  3,  3,  3,  3,  3,  3],
          [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  2,  2,  3,  3,  2,  2,
            2,  2,  2,  2,  3,  3,  3,  3,  3,  3,  3,  3,  3],
          [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  2,  2,  3,  2,  2,  2,
            2,  2,  2,  2,  3,  3,  3,  3,  3,  3,  3,  3,  3],
          [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,
            2,  2,  2,  2,  3,  3,  3,  3,  3,  3,  3,  3,  3],
          [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,
            2,  2,  2,  2,  2,  3,  3,  3,  3,  3,  3,  2,  2],
          [ 3,  3,  3,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,  2,  2,  2,
            2,  2,  2,  2,  2,  3,  3,  3,  3,  2,  2,  2,  2],
          [ 3,  3,  3,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,  2,  2,  2,
            2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2],
          [ 3,  3,  3,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,  2,  2,  2,
            2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2],
          [ 3,  3,  3,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,  2,  2,  2,
            2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2],
          [ 3,  3,  3,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,  2,  2,  2,
            2,  2,  2,  2,  2,  2,  2,  2,  4,  4,  2,  2,  2],
          [ 3,  3,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,
            2,  2,  2,  2,  2,  2,  2,  9,  9,  2,  2,  2,  2],
          [ 3,  3,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,
            2,  2,  2,  2,  2,  2,  3, 14,  8,  2,  2,  2,  2],
          [ 4,  4,  3,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,
            2,  2,  2,  2,  2,  2,  4, 15,  9,  2,  2,  2,  2],
          [ 5,  5,  4,  3,  3,  3,  3,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,
            2,  2,  2,  2,  2,  2,  4, 16,  9,  2,  2,  2,  2],
          [ 5,  5,  5,  4,  5,  5,  4,  3,  2,  2,  2,  2,  2,  2,  2,  2,  2,
            2,  2,  2,  2,  2,  2,  5, 17, 10,  2,  2,  2,  2],
          [ 5,  5,  5,  5,  5,  5,  5,  3,  2,  2,  2,  2,  2,  2,  2,  2,  2,
            2,  2,  2,  2,  2,  2, 10, 18, 10,  2,  2,  2,  2],
          [ 4,  4,  4,  4,  4,  3,  4,  2,  1,  1,  1,  1,  1,  1,  2,  2,  2,
            2,  2,  2,  2,  2,  4, 14, 18, 10,  2,  2,  2,  2],
          [ 1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  2,  2,
            2,  2,  2,  2,  2,  8, 18, 18, 12,  3,  2,  2,  2],
          [ 1,  1,  1,  1,  1,  1,  2,  2,  2,  2,  2,  2,  3,  3,  3,  4,  5,
            6,  6,  5,  5,  4, 12, 18, 18, 17, 13, 11, 10,  4],
          [ 6,  7,  8,  8,  8,  8,  8,  9,  9,  8,  8,  9,  9,  9, 10, 10, 10,
           10, 10, 10, 10, 10, 11, 13, 14, 16, 17, 17, 17, 14],
          [10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,
           10, 10, 10, 10, 10, 10, 10, 10, 10, 11, 11, 12, 12],
          [10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,
           10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10],
          [10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,
           10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,  9,  9],
          [ 9,  9,  9,  9,  9,  9,  9,  9,  9,  9,  9,  9,  9,  9,  9, 10, 10,
           10, 10, 10, 10, 10, 10,  9,  9,  8,  7,  7,  7,  6 ]]]])
           
           ```

ariG23498 commented:
Hi @capnmav77 now that the #37005 is merged, you could take a look at the segmentation maps logic. I think that could be of use to you.

capnmav77 commented:
hey @yonigozlan , the image processing fast segformer batched works perfectly but the labels of a single image are not matched , this discrepancy is likely due to inherent differences in the NEAREST interpolation between PIL (used by the slow processor) and TorchVision (used by the fast processor), which can be very difficult to perfectly align. what do you suggest i do ?

capnmav77 committed:
Merge remote-tracking branch 'upstream/main'

capnmav77 committed:
fixed some mismatched files

yonigozlan commented:
Hey @capnmav77 ! Sorry for the delay. Have you tried with NEAREST_EXACT? It could reduce the differences. If you still have issues we might just increase the tolerance for equivalence tests for this model

capnmav77 committed:
modified the tolerance for tests

simonreise commented:
Hi @capnmav77! I am not sure, but maybe applying the same fix as in #38042 may fix the issue with equivalence test. And it looks like this fix should be implemented anyway, because as far as I can see your current code applies `reduce_labels` not only to the labels, but also to the images

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into segformer-fast

yonigozlan committed:
use modular

yonigozlan reviewed (approved):
Hey @capnmav77 ! Sorry for the delay. Updated the PR to follow the last fast image processor standards and use modular, will merge once the CI passes!

yonigozlan committed:
fix ci

yonigozlan committed:
fixup

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into segformer-fast

yonigozlan committed:
fix modular
